warehouse/accounts/utils.py

```diff
@@ -0,0 +1,151 @@
+    # This is what we would expect to find if a basic auth value was leaked
+    "base64-basic-auth": Base64BasicAuthTokenMatcher(
+        # Basic auth is standard base64, so non-alphanumeric chars are + and /
+        # The following string was obtained by:
+        #     base64.b64encode(b"__token__:pypi-").decode("utf-8")
+        pattern=re.compile(r"X190b2tlbl9fOnB5cGkt[A-Za-z0-9+/=]+"),
+    ),
+}
+
+
+class InvalidTokenLeakRequest(Exception):
+    def __init__(self, message, reason):
+        self.reason = reason
+        super().__init__(message)
+
+
+@attr.s(auto_attribs=True, frozen=True, slots=True)
+class TokenLeakDisclosureRequest:
+
+    token: str
+    matcher: TokenMatcher
+    public_url: str
+
+    @classmethod
+    def from_api_record(cls, record, *, matchers=TOKEN_BREACH_MATCHERS):
+
+        if not isinstance(record, dict):
+            raise InvalidTokenLeakRequest(
+                f"Record is not a dict but: {type(record)}", reason="format"
+            )
+
+        try:
+            matcher_code = record["type"]
+        except KeyError:
+            raise InvalidTokenLeakRequest(
+                f"Record is missing attribute type", reason="format"
+            )
+
+        matcher = matchers.get(matcher_code)
+        if not matchers:
+            raise InvalidTokenLeakRequest(
+                f"Matcher with code {matcher_code} not found. "
+                f"Available codes are: {', '.join(matchers)}",
+                reason="invalid_matcher",
+            )
+        try:
+            return cls(token=record["token"], matcher=matcher, public_url=record["url"])
+        except KeyError as exc:
+            raise InvalidTokenLeakRequest(
+                f"Record is missing attribute {exc!s}", reason="format"
+            )
+
+
+class TokenLeakAnalyzer:
+    def __init__(self, request):
+        self._request = request
+
+    def analyze_disclosure(self, disclosure_record, origin):
+
+        metrics = self._request.find_service(IMetricsService, context=None)
+
+        metrics.increment(f"warehouse.token_leak.{origin}.recieved")
+
+        try:
+            disclosure = TokenLeakDisclosureRequest.from_api_record(
+                record=disclosure_record
+            )
+        except InvalidTokenLeakRequest as exc:
+            metrics.increment(f"warehouse.token_leak.{origin}.error.{exc.reason}")
+
+        # We don't expect to fail at this step: if GitHub starts sending us
+        # strings that don't match the regex we have, we have a problem.
+        # That being said, except logging, there's not much we can do.
+        try:
+            token = disclosure.matcher.extract(text=disclosure.token)
+        except ExtractionFailed:
+            metrics.increment(f"warehouse.token_leak.{origin}.error.extraction")
+            return
+
+        macaroon_service = self._request.find_service(IMacaroonService, context=None)
+
+        database_macaroon = macaroon_service.find_macaroon_from_raw(raw_macaroon=token)
```

Comment:
OMG ! I nearly missed that I need to somehow verify that the signature on that macaroon is correct, otherwise I'm just going to check that there is a macaroon with this ID in the DB.
It's probably not that bad, given I'd need to guess an existing UUID, but... Still...